### ISSUE TITLE
`bind_instead_of_map` should suggest removing `Result::and_then` call when possible

### ISSUE THREAD

**Issue by tgeoghegan** [rust-lang/rust-clippy#14801]
### Summary

`bind_instead_of_map` transforms certain `Result::and_then` calls into `Result::map`, but in the example I provide below, it'd be better to just remove `and_then` altogether. Sure enough, if I apply clippy's suggestion to replace `and_then(|v| Ok(v)) with `map(|v| v)`, the fixed code then triggers `map_identity`, which tells me to remove the `map` call clippy just bullied me into adding.

### Reproducer

I tried this code:

```rust
fn returns_result() -> Result<(), ()> {
    Ok(())
}

fn main() {
    returns_result().and_then(|v| Ok(v)).unwrap();
}
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=0791274bab86244698fe5256afb33e1c

I expected to see this happen:

Clippy should suggest removing the `and_then` altogether, yielding `returns_result().unwrap();`

Instead, this happened:

Clippy suggests using `map`:

```
warning: using `Result.and_then(|x| Ok(y))`, which is more succinctly expressed as `map(|x| y)`
 --> src/main.rs:6:5
  |
6 |     returns_result().and_then(|v| Ok(v)).unwrap();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `returns_result().map(|v| v)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bind_instead_of_map
  = note: `#[warn(clippy::bind_instead_of_map)]` on by default
```

...which triggers `map_identity`:

```
warning: unnecessary map of the identity function
 --> src/main.rs:6:21
  |
6 |     returns_result().map(|v| v).unwrap();
  |                     ^^^^^^^^^^^ help: remove the call to `map`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_identity
  = note: `#[warn(clippy::map_identity)]` on by default
```

### Version

```text
Reproduced on play.rust-lang.org today, which is currently Rust 1.86, Clippy 0.1.86
```

**Comment by samueltardieu** [comment]
Clippy is supposed to be used this way: a lint may suggest a better rewrite, which in turn will trigger another lint which will trigger an even better rewrite. Special-casing all lint suggestions to ensure that Clippy never suggests fixes which could themselves trigger other lints will be a daunting task.

**Comment by tgeoghegan** [comment]
Oh, I wouldn't want clippy to ignore this. Rather, I would like it to skip the first suggestion and go right to the second. The scenario I'm concerned about is running clippy in continuous integration: imagine I put out a PR, it fails due to this lint, I fix the lint and re-submit -- and find it still fails with another lint. Frustrating, especially if your CI happens to take a long time (sadly common for Rust projects).

Naively, without understanding clippy or rustc internals, I might suggest that clippy could re-run itself until it stops making suggestions, and then emit a chain of suggested fixes to the user. But if the maintainers feel that isn't worth the effort, I won't object to closing this as "behaves correctly".

**Comment by Jarcho** [comment]
`clippy --fix` actually does do what you suggest. Doing so on regular clippy runs is not really feasable since it requires changing files on disk to rerun the compiler.
